### PR TITLE
GH-2443: Get auction info backwards compatibility

### LINKED ISSUE #2443: The casper-client cannot retrieve get-auction-info for blocks from version 1.3.4 and before.
FWIW - I tried this both with casper-client v1.3.2  and casper-client v1.4.2 - so it’s not a backward compatibility issue in a newer client. The older client also wasn’t able to see further back than 1.4.1

Reporter Michael / Make

### PR DESCRIPTION
**Pull request by mpapierski**
Closes #2443

This PR fixes a problem of the get-auction-state query for pre #1814 (introduction of `Key::SystemContractRegistry`) blocks by passing down an instance of `SystemContractRegistry` initialized from a highest block's state root hash at component initialization time. This avoids querying the registry from a specified state which may not exist for older blocks.

### CODE REVIEW

**Comment by goral09** [comment]
@mpapierski What's your plan to verify that this fixes the problem?

**Comment by mpapierski** [comment]
> @mpapierski What's your plan to verify that this fixes the problem?

An integration network and casper-client quering get-auction-info through RPC for 1.3.x state root hashes seems like easiest approach

**Review by goral09 — APPROVED** [review]

**Comment by goral09** [comment]
> > @mpapierski What's your plan to verify that this fixes the problem?
> 
> An integration network and casper-client quering get-auction-info through RPC for 1.3.x state root hashes seems like easiest approach

Is that going to be a separate PR?

**Review by darthsiroftardis — COMMENTED** [review]
The code seems fine, but I think we should have a test to see that the fix works before we merge this in

File: node/src/components/contract_runtime.rs
Comment (Fraser999) [inline node/src/components/contract_runtime.rs:142]: Please add a comment as to why we need this member variable, since it's not obvious.

File: node/src/components/contract_runtime/operations.rs
Comment (Fraser999) [inline node/src/components/contract_runtime/operations.rs:37]: I don't think we should be passing `system_contract_registry` here.  In the `feat-fast-sync` branch, we only call `contract_runtime.set_initial_state` in the participating reactor, but I believe `execute_finalized_block` can be called from the chain synchronizer in the joiner reactor.  In that case, it would have an empty `SystemContractRegistry`.

We should probably change `get_era_validators` to take an optional `SystemContractRegistry`, and if it's `None` try to retrieve from the tracking copy.

**Review by Fraser999 — CHANGES_REQUESTED** [review]
We need a changelog entry (probably just in node) for the fix.

Blocking the PR for now until we can discuss my second comment.

File: node/src/components/contract_runtime/operations.rs
Comment (goral09) [reply node/src/components/contract_runtime/operations.rs:37]: > In the feat-fast-sync branch, we only call `contract_runtime.set_initial_state` in the participating reactor, but I believe` execute_finalized_block` can be called from the chain synchronizer in the joiner reactor. In that case, it would have an empty `SystemContractRegistry`.

This PR is made to `dev`, not `feat-fast-sync` though. Making `SystemContractRegistry` optional in the `ContractRuntime` component will add unnecessary work to the PR, code that won't quite make sense for `dev`. We can make that work when migrating this fix to `feat-fast-sync`. 

Also, you're right that it optional, and reading from the `scratch_global_state`, is a solution.

**Pull request closed without merging**
